File: src/main/java/domain/fan/ScotchPatience.java
package domain.fan;

import domain.Constraint;
import domain.constraints.AndConstraint;
import domain.constraints.MoveInformation;
import domain.constraints.NextRank;
import domain.constraints.OppositeColor;
import domain.constraints.movetypes.MoveComponents;
import domain.constraints.movetypes.TopCardOf;

public class ScotchPatience extends Domain{

    public ScotchPatience(){
        super("ScotchPatience");
        init();
    }
    /**
     * Determines what cards can be placed on tableau.
     *
     * Parameter is either a single card
     *
     * @param bottom
     * @return
     */
    @Override
    public Constraint buildOnTableau(MoveInformation bottom) {
        TopCardOf topDestination = new TopCardOf(MoveComponents.Destination);
        return new NextRank(topDestination, bottom);
    }

    /**
     * Determines what cards can be placed on tableau.
     *
     * Parameter is either a single card, or something like BottomOf().
     *
     * @param bottom
     * @return
     */
    @Override
    public Constraint buildOnFoundation(MoveInformation bottom) {
        TopCardOf topDestination = new TopCardOf(MoveComponents.Destination);
        return new AndConstraint(new NextRank(bottom, topDestination), new OppositeColor(bottom, topDestination));
    }

}

File: src/main/scala/org/combinators/solitaire/fan/scotchpatience.scala
package org.combinators.solitaire.fan

import org.combinators.solitaire.domain._

package object scotchpatience extends variationPoints {

  override def buildOnTableau (card: MovingCard.type): Constraint = {
    val topDestination = TopCardOf(Destination)
    NextRank(topDestination, card)   // suit not relevant...
  }

  override def buildOnFoundation (card: MovingCard.type): Constraint = {
    val topDestination = TopCardOf(Destination)
    AndConstraint(NextRank(card, topDestination),  OppositeColor(card, topDestination)) //can be the opposite color
  }

//  /**
//    * Clear 0th Foundation and place [Two, Ace] on the 0th Tableau
//    */
//  case object PrepareTableauToFoundation extends Setup {
//    val sourceElement = ElementInContainer(Tableau, 0)
//    val targetElement = Some(ElementInContainer(Foundation, 2))
//
//    // clear Foundation, and place [2C, AC] on 0th tableau
//    val setup:Seq[SetupStep] = Seq(
//      RemoveStep(ElementInContainer(Tableau, 1)),
//      RemoveStep(ElementInContainer(Tableau, 2)),
//      InitializeStep(ElementInContainer(Tableau, 2), CardCreate(Clubs, Ace)),
//      InitializeStep(ElementInContainer(Tableau, 2), CardCreate(Hearts, Two)),
//      InitializeStep(ElementInContainer(Foundation, 2), CardCreate(Spades, Ace)),
//      InitializeStep(ElementInContainer(Foundation, 2), CardCreate(Diamonds, Two)),
//    )
//  }

  case object TableauToNextTableauIgnoreSuit extends Setup {
    val sourceElement = ElementInContainer(Tableau, 1)
    val targetElement = Some(ElementInContainer(Tableau, 2))

    override val setup:Seq[SetupStep] = Seq(
      RemoveStep(sourceElement),
      RemoveStep(targetElement.get),
      InitializeStep(targetElement.get, CardCreate(Clubs, Three)),
      MovingCardStep(CardCreate(Hearts, Two))
    )
  }
  case object TableauToNextFoundationOppositeColor extends Setup {
    val sourceElement = ElementInContainer(Tableau, 1)
    val targetElement = Some(ElementInContainer(Foundation, 2))

    override val setup:Seq[SetupStep] = Seq(
      RemoveStep(sourceElement),
      RemoveStep(targetElement.get),
      InitializeStep(targetElement.get, CardCreate(Clubs, Ace)),
      MovingCardStep(CardCreate(Hearts, Two))
    )
  }

  val scotchpatience:Solitaire = {
    Solitaire(name = "ScotchPatience",
      structure = structureMap,
      layout = Layout(layoutMap),
      deal = getDeal,
      specializedElements = Seq.empty,
      moves = Seq(tableauToTableauMove, tableauToFoundationMove),
      logic = BoardState(Map(Tableau -> 0, Foundation -> 52)),
      solvable = true,
      customizedSetup = Seq(//PrepareTableauToFoundation,
        TableauToNextTableauIgnoreSuit, TableauToNextFoundationOppositeColor,
        TableauToEmptyFoundation, TableauToNextFoundation, //shouldn't pass - has to be opposite color
        TableauToEmptyTableau, TableauToNextTableau)
    )
  }
}
